crates/cext/src/dag.rs

```diff
@@ -157,6 +163,509 @@ pub unsafe extern "C" fn qk_dag_num_clbits(dag: *const DAGCircuit) -> u32 {
+///     QkQuantumRegister *qr = qk_quantum_register_new(1, "my_register");
+///     qk_dag_add_quantum_register(dag, qr);
+///
+///     uint32_t qubit[1] = {0};
+///     qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+///     size_t num = qk_dag_num_op_nodes(dag); // 1
+///
+///     qk_dag_free(dag);
+///     qk_quantum_register_free(qr);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_num_op_nodes(dag: *const DAGCircuit) -> usize {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    dag.num_ops()
+}
+
+/// @ingroup QkDag
+///
+/// DAG node type.
+#[repr(u8)]
+pub enum CDagNodeType {
+    /// Operation node.
+    Operation = 0,
+    /// Qubit wire start node.
+    QubitIn = 1,
+    /// Qubit wire end node.
+    QubitOut = 2,
+    /// Clbit wire start node.
+    ClbitIn = 3,
+    /// Clbit wire end node.
+    ClbitOut = 4,
+    /// Classical variable wire start node.
+    VarIn = 5,
+    /// Classical variable wire end node.
+    VarOut = 6,
+}
+
+/// @ingroup QkDag
+/// Get the type of the specified node.
+///
+/// The result can be used in a switch statement to dispatch proper handling
+/// when iterating over nodes of unknown type.
+///
+/// @param dag A pointer to the DAG.
+/// @param qubit The node to get the type of.
+///
+/// @return The type of the node.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_node_type(dag: *const DAGCircuit, node: u32) -> CDagNodeType {
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    match dag
+        .dag()
+        .node_weight(NodeIndex::new(node as usize))
+        .unwrap()
+    {
+        NodeType::QubitIn(_) => CDagNodeType::QubitIn,
+        NodeType::QubitOut(_) => CDagNodeType::QubitOut,
+        NodeType::ClbitIn(_) => CDagNodeType::ClbitIn,
+        NodeType::ClbitOut(_) => CDagNodeType::ClbitOut,
+        NodeType::VarIn(_) => CDagNodeType::VarIn,
+        NodeType::VarOut(_) => CDagNodeType::VarOut,
+        NodeType::Operation(_) => CDagNodeType::Operation,
+    }
+}
+
+/// @ingroup QkDag
+/// Retrieve the index of the input node of the wire corresponding to the given qubit.
+///
+/// @param dag A pointer to the DAG.
+/// @param qubit The qubit to get the input node index of.
+///
+/// @return The input node of the qubit wire.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_qubit_in_node(dag: *const DAGCircuit, qubit: u32) -> u32 {
+    let dag = unsafe { const_ptr_as_ref(dag) };
+    dag.qubit_io_map()[qubit as usize][0].index() as u32
+}
```

Comment:
Not saying I prefer one way or the other, just raising the question:

We expose `qubit_io_map` as `&[[NodeIndex; 2]]` in Rust - did we consider `qk_dag_qubit_endpoints(dag: *const DAGCircuit) -> *const u32` and document how C's supposed to interact with it (i.e. it's a strided pointer)?

Reply:
I remember we talked about it briefly and there was some concern about lifetimes in the event that more wires are added and the `Vec` reallocates and thus invalidates the old data pointer.

I went for this approach specifically to avoid the complexity of that for users.